lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -3411,6 +3451,28 @@ static void setFP16AccumSLS(Function *F,
   } while (nodeIt != stopIt);
 }
 
+static void transformIndexTypeDemotion(const Backend &B, Function *F,
```

Comment:
OK. Now that I look at this again, it is indeed generic as it relies on knowledge of the type's actual bitwidth. At least in its current state it does not differentiate the semantics of that integer data: it can be whatever as long as it's safe to clip the data without losing semantical information (actualSize()).